vault/resource_approle_auth_backend_role.go

```diff
@@ -0,0 +1,384 @@
+	}
+
+	d.Set("backend", backend)
+	d.Set("role_name", role)
+	d.Set("period", period)
+	d.Set("policies", policies)
+	d.Set("bound_cidr_list", cidrs)
+	d.Set("secret_id_num_uses", secretIDNumUses)
+	d.Set("secret_id_ttl", secretIDTTL)
+	d.Set("token_num_uses", tokenNumUses)
+	d.Set("token_ttl", tokenTTL)
+	d.Set("token_max_ttl", tokenMaxTTL)
+	d.Set("period", period)
+	d.Set("bind_secret_id", resp.Data["bind_secret_id"])
+
+	log.Printf("[DEBUG] Reading AppRole auth backend role %q RoleID", path)
+	resp, err = client.Logical().Read(path + "/role-id")
+	if err != nil {
+		return fmt.Errorf("Error reading AppRole auth backend role %q RoleID: %s", path, err)
+	}
+	log.Printf("[DEBUG] Read AppRole auth backend role %q RoleID", path)
+	if resp != nil {
+		d.Set("role_id", resp.Data["role_id"])
+	}
+
+	return nil
+}
+
+func approleAuthBackendRoleUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Id()
+
+	log.Printf("[DEBUG] Updating AppRole auth backend role %q", path)
+	iPolicies := d.Get("policies").([]interface{})
+	policies := make([]string, 0, len(iPolicies))
+	for _, iPolicy := range iPolicies {
+		policies = append(policies, iPolicy.(string))
+	}
+
+	iCIDRs := d.Get("bound_cidr_list").([]interface{})
+	cidrs := make([]string, 0, len(iCIDRs))
+	for _, iCIDR := range iCIDRs {
+		cidrs = append(cidrs, iCIDR.(string))
+	}
+
+	data := map[string]interface{}{
+		"policies":           policies,
+		"bound_cidr_list":    strings.Join(cidrs, ","),
+		"bind_secret_id":     d.Get("bind_secret_id").(bool),
+		"secret_id_num_uses": d.Get("secret_id_num_uses").(int),
+		"secret_id_ttl":      d.Get("secret_id_ttl").(int),
+		"token_num_uses":     d.Get("token_num_uses").(int),
+		"token_ttl":          d.Get("token_ttl").(int),
+		"token_max_ttl":      d.Get("token_max_ttl").(int),
+		"period":             d.Get("period").(int),
+	}
+
+	_, err := client.Logical().Write(path, data)
+
+	d.SetId(path)
+
+	if err != nil {
+		return fmt.Errorf("Error updating AppRole auth backend role %q: %s", path, err)
+	}
+	log.Printf("[DEBUG] Updated AppRole auth backend role %q", path)
+
+	if d.HasChange("role_id") {
+		log.Printf("[DEBUG] Updating AppRole auth backend role %q RoleID", path)
+		_, err := client.Logical().Write(path+"/role-id", map[string]interface{}{
+			"role_id": d.Get("role_id").(string),
+		})
+		if err != nil {
+			return fmt.Errorf("Error updating AppRole auth backend role %q's RoleID: %s", path, err)
+		}
+		log.Printf("[DEBUG] Updated AppRole auth backend role %q RoleID", path)
+	}
+
+	return approleAuthBackendRoleRead(d, meta)
+
+}
+
+func approleAuthBackendRoleDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*api.Client)
+	path := d.Id()
+
+	log.Printf("[DEBUG] Deleting AppRole auth backend role %q", path)
+	_, err := client.Logical().Delete(path)
+	if err != nil {
+		return fmt.Errorf("Error deleting AppRole auth backend role %q", path)
```

Comment:
Q: Is it possible that `Delete` here returns some kind of error akin to a 404? If so, we should catch that error and simply unset the ID

Reply:
I could've _sworn_ that Vault just returned nil on 404, instead of treating it as an error, but upon looking to prove that, discovered it _will_ treat a 404 as an error, but in a way that's incredibly frustrating to catch. :(